woocommerce/grow
# Add Compat Checker library for versioning and compatibility checks

### Changes proposed in this Pull Request:


Closes #75.

This PR:

* Adds a new `PHP` package called `compat-checker` that can be installed via composer in WooCommerce Grow extensions.
* This new package runs compatibility checks based on the plugin's header file.
* Here is a list of compatibility checks run:
    * Checks if the WooCommerce plugin is installed and active.
    * Checks if the active WooCommerce version meets the minimum WooCommerce version defined in the plugin header.
    * Displays a warning if the active WooCommerce version is behind L-N (2 is the default) version.
    * Displays a warning if the active WooCommerce is untested with the plugin.
    * Displays a warning if the current WordPress version is untested with the plugin.


### Detailed test instructions:


#### Setting up the fork.

1. Open the WooCommerce Grow repository from the URL: https://github.com/woocommerce/grow
2. On the top right, click the **Fork** button to fork your own copy.
3. Ensure the `Copy the trunk branch only` is unchecked.
4. Click the `Create Fork` button. This step will create a fork of [`woocommerce/grow`](https://github.com/woocommerce/grow). Let's say: [`ibndawood/grow`](https://github.com/ibndawood/grow/)
5. In the forked repository, create a new PR to merge the branch `add/compat-checker` branch into `trunk`.
6. Merge the PR.

#### Testing the GitHub Workflow

1. Navigate to `Actions` tab in the forked repository. 
2. Ensure `Public Compat Checker package` workflow is added.
3. Ensure there is a branch named `compat-checker`.
4. Make a simple commit to the `trunk` to trigger the workflow.
5. Ensure the subdirectory `packages/php/compat-checker` contents are published into the `compat-checker` branch.

#### Including the library in the WooCommerce Extension

1. Clone any one of the WooCommerce Grow extensions. Let's say: [WooCommerce Brands](https://github.com/woocommerce/woocommerce-brands/)
2. Include the `compat-checker` package by adding the repository to the `composer.json` of the cloned repository like in the snippet below. Replace `https://github.com/ibndawood/grow` with the URL of your fork:
```json
"repositories": [
    {
        "type": "vcs",
        "url": "https://github.com/ibndawood/grow"
    }
],
"require": {
    "woocommerce/grow": "dev-compat-checker"
},
```
3. Run `composer update` to included `woocommerce/grow` in the plugin's `vendor` folder.
4. Ensure the `vendor` folder has the `woocommerce/grow` folder.
5. Ensure the `vendor/composer/autoload_psr4.php` file has `Automattic\WooCommerce\Grow\Tools\` namespace included.

#### Testing compatibility and versioning

##### Setup
In the main plugin file, run `Automattic\WooCommerce\Grow\Tools\CompatChecker\v0_0_1\Checker::instance()->is_compatible( __FILE__, $plugin_version )` to check for compatibility and versioning. The snippet should be placed before initialising the plugin. The below example snippet is for WooCommerce Brands:
```php
require __DIR__ . '/vendor/autoload.php';
use Automattic\WooCommerce\Grow\Tools\CompatChecker\v0_0_1\Checker;

// Plugin init hook.
add_action( 'plugins_loaded', 'wc_brands_init', 1 );

/**
 * Initialize plugin.
 */
function wc_brands_init() {
    define( 'WC_BRANDS_VERSION', '1.6.56' ); // WRCS: DEFINED_VERSION.

    if ( ! Checker::instance()->is_compatible( __FILE__, WC_BRANDS_VERSION ) ) {
	return;
    }
    // Continue with initializing...
}
```
##### WooCommerce activation tests

With the above code snippets in place, run the following tests. The instructions are for WooCommerce Brands but can be used for other WooGrow Extensions.

1. Install and activate WooCommerce Brands without installing WooCommerce.
2. Deactivate WooCommerce.
3. Verify the notice `WooCommerce Brands requires WooCommerce to be installed and activated. Please install WooCommerce.` is displayed.
4. Install WooCommerce without activating.
5. Verify the notice `WooCommerce Brands requires WooCommerce to be activated. Please activate WooCommerce.` is displayed.

##### WooCommerce versioning tests
⚠️  **Important note**: You will need to erase the transient `_transient_wc_grow_compat_checker_[EXTENSION].php[VERSION` after each change!!

1. Install and activate WooCommerce.
2. Install and activate WooCommerce Brands.
3. Modify the `WC requires at least` header field and set it to a version higher than the active WooCommerce version. Let's say the `WC requires at least` is set to `8.2` and the active WooCommerce version is `8.0.2`.
4. Verify the notice `WooCommerce Brands requires WooCommerce version 8.2 or higher. Please update WooCommerce to the latest version, or download the minimum required version` is displayed.
5. Modify the `WC tested up to` and set it to a lower version than the active WooCommerce version. Let's say the `WC tested up to` is set to `7.8`, the active WooCommerce version is `8.0.2` and WooCommerce Brands version is `1.6.56`.
6. Verify the warning notice `WooCommerce Brands - 1.6.56 is untested with WooCommerce 8.0.2` is displayed.

##### L2 support recommendation notice

1. Download the latest minus three WooCommerce version and activate it. Let's say `7.6`.
2. Verify the warning notice `Heads up! WooCommerce Brands will soon discontinue support for WooCommerce 7.6. Please update WooCommerce to take advantage of the latest updates and features.`

##### WordPress versioning tests
⚠️  **Important note**: You will need to erase the transient `_transient_wc_grow_compat_checker_[EXTENSION].php[VERSION` after each change!!

1. Modify the `Tested up to` and set it to a lower version than the current WordPress version. Let's say the `Tested up to` is set to `6.2`, the active WordPress version is `6.3`, and the WooCommerce Brands version is `1.6.56`.
2. Verify the warning notice `WooCommerce Brands - 1.6.56 is untested with WordPress 6.3` is displayed.


### Changelog entry

> Add - Compat Checker library for versioning and compatibility checks.
